File: src/commonMain/kotlin/world/ContainerManager.kt
package world

import inventory.ItemStack
import inventory.LootTables

/**
 * Manages lootable containers on the map.
 * Loot is generated on first open, then persisted.
 */
class ContainerManager(private val world: GameWorld) {
    data class ContainerData(
        val tileX: Int,
        val tileY: Int,
        val type: String,
        val items: MutableList<ItemStack>,
        var opened: Boolean = false
    ) {
        val displayName: String get() = localization.Lang["container.name.$type"]
    }

    private val containers = mutableMapOf<Long, ContainerData>()

    init {
        // Scan entire map for any tile with a containerType
        for (y in 0 until GameWorld.MAP_HEIGHT) {
            for (x in 0 until GameWorld.MAP_WIDTH) {
                val tile = world.getTile(x, y)
                if (tile != null && tile.containerType != null) {
                    val key = posKey(x, y)
                    containers[key] = ContainerData(x, y, tile.containerType, mutableListOf())
                }
            }
        }
    }

    private fun posKey(x: Int, y: Int): Long = y.toLong() * GameWorld.MAP_WIDTH + x

    fun findNearby(px: Double, py: Double, range: Double = 48.0): ContainerData? {
        val ts = GameWorld.TILE_SIZE
        for (container in containers.values) {
            val cx = container.tileX * ts + ts / 2.0
            val cy = container.tileY * ts + ts / 2.0
            val dx = px - cx
            val dy = py - cy
            if (dx * dx + dy * dy <= range * range) {
                return container
            }
        }
        return null
    }

    fun open(container: ContainerData): List<ItemStack> {
        if (!container.opened) {
            container.items.addAll(LootTables.generate(container.type))
            container.opened = true
        }
        return container.items
    }

    fun takeItem(container: ContainerData, index: Int): ItemStack? {
        if (index < 0 || index >= container.items.size) return null
        val stack = container.items[index]
        if (stack.count > 1) {
            stack.count--
            return ItemStack(stack.def, 1)
        } else {
            return container.items.removeAt(index)
        }
    }
}

File: scripts/world/ContainerManager.cs
using System;
using System.Collections.Generic;
using Godot;
using ProiectDeLegume.Scripts.Inventory;
using ProiectDeLegume.Scripts.Localization;

namespace ProiectDeLegume.Scripts.World;

public class ContainerData
{
    public int TileX { get; }
    public int TileY { get; }
    public string Type { get; }
    public List<ItemStack> Items { get; } = new();
    public bool Opened { get; set; }

    public ContainerData(int tileX, int tileY, string type)
    {
        TileX = tileX;
        TileY = tileY;
        Type = type;
    }

    public string DisplayName => Lang.Get($"container.name.{Type}");
}

public class ContainerManager
{
    private readonly Dictionary<long, ContainerData> _containers = new();
    private const int TileSize = TileSetGenerator.TileSize;

    // Container tile indices and their loot table type
    private static readonly Dictionary<int, string> ContainerTiles = new()
    {
        { TileSetGenerator.TrashCan, "trash_can" },
        { TileSetGenerator.Wardrobe, "wardrobe" },
        { TileSetGenerator.KitchenCabinet, "kitchen" },
        { TileSetGenerator.Medkit, "medkit" },
        { TileSetGenerator.GarageShelf, "garage" },
    };

    public void ScanMap(TileMapLayer objectLayer)
    {
        for (int y = 0; y < MapGenerator.MapHeight; y++)
        {
            for (int x = 0; x < MapGenerator.MapWidth; x++)
            {
                var coords = objectLayer.GetCellAtlasCoords(new Vector2I(x, y));
                if (coords.X >= 0 && ContainerTiles.TryGetValue(coords.X, out var type))
                {
                    _containers[PosKey(x, y)] = new ContainerData(x, y, type);
                }
            }
        }

        GD.Print($"Found {_containers.Count} containers");
    }

    private long PosKey(int x, int y) => (long)y * MapGenerator.MapWidth + x;

    public ContainerData FindNearby(double px, double py, double range = 48.0)
    {
        double rangeSq = range * range;
        foreach (var container in _containers.Values)
        {
            double cx = container.TileX * TileSize + TileSize / 2.0;
            double cy = container.TileY * TileSize + TileSize / 2.0;
            double dx = px - cx;
            double dy = py - cy;
            if (dx * dx + dy * dy <= rangeSq)
                return container;
        }
        return null;
    }

    public List<ItemStack> Open(ContainerData container)
    {
        if (!container.Opened)
        {
            container.Items.AddRange(LootTables.Generate(container.Type));
            container.Opened = true;
        }
        return container.Items;
    }
}
